src/rustc.rs

```diff
@@ -19,20 +20,212 @@ impl TargetList {
 }
 
 pub trait VersionMetaExt {
-    fn host(&self) -> Host;
+    fn host(&self) -> TargetTriple;
     fn needs_interpreter(&self) -> bool;
 }
 
 impl VersionMetaExt for VersionMeta {
-    fn host(&self) -> Host {
-        Host::from(&*self.host)
+    fn host(&self) -> TargetTriple {
+        TargetTriple::from(&*self.host)
     }
 
     fn needs_interpreter(&self) -> bool {
         self.semver < Version::new(1, 19, 0)
     }
 }
 
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+pub struct QualifiedToolchain {
+    pub channel: String,
+    pub date: Option<String>,
+    pub(self) host: TargetTriple,
+    pub is_custom: bool,
+    pub full: String,
+}
+
+impl QualifiedToolchain {
+    pub fn new(channel: &str, date: &Option<String>, host: &TargetTriple) -> Self {
+        Self {
+            channel: channel.to_string(),
+            date: date.clone(),
+            host: host.clone(),
+            is_custom: false,
+            full: if let Some(date) = date {
+                format!("{}-{}-{}", channel, date, host)
+            } else {
+                format!("{}-{}", channel, host)
+            },
+        }
+    }
+
+    /// Replace the host, does nothing if ran on a custom toolchain
+    pub fn replace_host(&mut self, host: &TargetTriple) -> &mut Self {
+        if !self.is_custom {
+            *self = Self::new(&self.channel, &self.date, host);
+        }
+        self
+    }
+
+    /// Makes a good guess as to what the toolchain is compiled to run on.
+    pub(crate) fn custom(name: &str, sysroot: &Path, verbose: bool) -> Result<QualifiedToolchain> {
+        if let Ok(compat) = std::env::var("CROSS_CUSTOM_TOOLCHAIN_COMPAT") {
+            let mut toolchain: QualifiedToolchain = compat.parse().wrap_err(
+                "could not parse CROSS_CUSTOM_TOOLCHAIN_COMPAT as a fully qualified toolchain name",
+            )?;
+            toolchain.is_custom = true;
+            toolchain.full = name.to_string();
+            return Ok(toolchain);
+        }
```

Comment:
new feature in conjunction with #803 , this is because we now need to know what host this toolchain is supposed to run on